### ISSUE TITLE
TorchDistributedTrial properties can make torch.distributed hang

### ISSUE THREAD

**Issue by nlgranger** [optuna/optuna#3484]
### Expected behavior

Reading TorchDistributedTrial properties should have no side effects, in particular it should not call torch.distributed methods. 

```
# ...
if rank == 0:
    print(hp_search_trial.number)

torch.distributed.barrier()
# execution continues
```

### Environment

- Optuna version: 2.10.0
- Python version: 3.9.12
- OS: Linux-4.18.0-240.el8.x86_64-x86_64-with-glibc2.28
- (Optional) Other libraries and their versions: Pytorch 1.11

### Error messages, stack traces, or logs

Processes hang after a trial property is read on one node only.
In the example above, reading the trial number on only one node causes a call to a torch.distributed method on one node only. The next distributed functions will hang because it expects the other nodes to call the same method first.

### Steps to reproduce

```py
import torch
import optuna
from torch import multiprocessing


def train(rank, world_size, trial):
    print(f"process rank {rank} started.")

    torch.distributed.init_process_group(
        backend="gloo",
        init_method="tcp://127.0.0.1:29400",
        world_size=world_size,
        rank=rank)

    print(f"process rank {rank} initialized.")

    trial = optuna.integration.TorchDistributedTrial(trial if rank == 0 else None)

    torch.distributed.barrier()
    print(f"process rank {rank} ready.")

    if rank == 0:
        print(trial.number)  # <<<------ rank 0 makes distributed op alone

    torch.distributed.barrier()  # <<<------ will hang here because rank 1 is late by one distributed op
    print(f"process rank {rank} exiting.")


def run_trial(trial):
    multiprocessing.spawn(train, (2, trial), nprocs=2)
    return 0


def main():
    study = optuna.create_study()
    study.optimize(run_trial, n_trials=1)


if __name__ == "__main__":
    main()
```

### Additional context (optional)

torch.distributed essentially works by keeping a history of called functions. For instance on two nodes, it may look like this:

| index | rank0 | rank1 |
|--|--|--|
| 0 | dist.all_reduce | dist.all_reduce |
| 1 | dist.all_reduce | dist.all_reduce |
| 2 | dist.broadcast |  |

In this example rank1 is a bit late so rank0 will hang before executing the next distributed operation.

Back to optuna, a common use-case in a torch.distributed setup is to log or save a checkpoint on rank 0:

```
if torch.distributed.get_rank() == 0:
    print(hp_search_trial.number)

torch.distributed.barrier()
# execution continues
```

Before this code, all nodes participating in the distributed environment have issued the same distributed functions calls.
The current implementation of `hp_search_trial.number` causes a call to torch.distributed.broadcast **but only on rank 0**.
Upon reaching the next distributed operation (`torch.distributed.barrier()` in this example), pytorch hangs on rank0 until rank1 executes a broadcast to mirror rank0 operations but this never happens.

Optuna documentation already states that trial methods use the distributed api and thus must be called on all nodes, but properties are not mentioned.
I think it is inconvenient and non-pythonic that reading properties has a side effect.

**Comment by nzw0301** [comment]
I'm sorry I could not understand what is issue. Could you elaborate a little bit more about the bug?  We really appreciate you if you share minimal reproducible codes with us.

**Comment by nlgranger** [comment]
Sure! I'm working on it and a PR too.

**Comment by nzw0301** [comment]
Amazing. Thanks!

**Comment by nlgranger** [comment]
I have updated the issue with an MRE and submitted the PR.

**Comment by toshihikoyanase** [comment]
I'm afraid that this is not a bug, but an expected behavior of the current `TorchDistriubtedTrial`.

[The official example](https://github.com/optuna/optuna-examples/blob/e3ea68dece4bf2f59a5d1731139e13b416aa9cb2/pytorch/pytorch_distributed_simple.py#L99-L100) distinguishes `single_trial: Optional[Trial]` and `trial: TorchDistributedTrial`. Note that `single_trial` is an instance of `Trial` if rank==0, otherwise, it is `None`.

The rank-0 node has a Trial instance, so the code should refer to it there.
```
    if rank == 0:
        assert single_trial is not None
        print(single_trial.number)  # <<<------ Use the  Trial instance instead of `TorchDistributedTrial`
```

Currently, users need to choose single trials and distributed trials carefully, and it may be annoying. So, I appreciate your PR that will improve usability.

**Comment by toshihikoyanase** [comment]
Let me change the label to `enhancement`.

**Comment by nlgranger** [comment]
The doc does not mention attributes actually being properties with calls to torch.distributed.broadcast, that got me confused.

Your comment makes me think we could simply remove the proxy to the single trial properties when rank != 0 and return the readily available single trial properties when rank == 0. It would be a lot simpler than the proposed PR and would likely not impact many users. Are there any other optuna functions that would need access to the trial properties on rank != 0 ?

**Comment by toshihikoyanase** [comment]
> simply remove the proxy to the single trial properties when rank != 0 and return the readily available single trial properties when rank == 0.

Thank you for your proposal. I couldn't imagine the implementation, so could you share some code examples?
More specifically, I think a `TorchDistributedTrial` instance in rank-0 node needs to broadcast properties in order for the other rank nodes to get the value of properties.
